core/arch/aarch64/proc.c

```diff
@@ -88,11 +88,35 @@ get_processor_specific_info(void)
     cpu_info.features.flags_aa64mmfr1 = isa_features[AA64MMFR1];
     cpu_info.features.flags_aa64dfr0 = isa_features[AA64DFR0];
 
-#        if !defined(DR_HOST_NOT_TARGET) && defined(SVE)
-    /* TODO i#3044: Vector length will be set by reading value from h/w. */
-    CLIENT_ASSERT(false, "TODO i#3044: SVE requires initialisation of vector length!");
-#        elif !defined(STANDALONE_DECODER) || defined(DR_HOST_NOT_TARGET)
-    /* Set SVE vector length for unit tests. */
+    /* The SVE vector length is set to:
+     * - A value read from the host hardware.
+     * or:
+     * - 32 bytes, 256 bits.
+     * Which of the above depends on:
+     * - SVE or non-SVE AArch64 or x86 host h/w.
+     * and:
+     * - Release or development test build.
+     */
+#        if !defined(DR_HOST_NOT_TARGET)
+    if (proc_has_feature(FEATURE_SVE)) {
+#            if !defined(BUILD_TESTS)
+        uint64 vl;
+        /* TODO i#5365: Ideally this should be generated by INSTR_CREATE_rdvl()
+         * and executed at startup time with other initialisation code.
+         */
+        asm(".inst 0x04bf5020\n" /* rdvl x0, #1 */
+            "mov %0, x0"
+            : "=r"(vl)
+            :
+            : "x0");
+        cpu_info.sve_vector_length_bytes = vl;
+#            else
+        cpu_info.sve_vector_length_bytes = 32;
```

Comment:
The test reference data for SVE instructions is static and assumes a VL of 256 bits, e.g.
`e58057a1 : str z1, [x29, #5, mul vl]                : str    %z1 -> +0x05(%x29)[32byte]`
in https://raw.githubusercontent.com/DynamoRIO/dynamorio/master/suite/tests/api/dis-a64-sve.txt
So, when we build and test on SVE and non-SVE the VL is fixed to 256 bits regardless.

Reply:
(Sorry for the delay; lost track of this.). But we also need to test the user dynamically setting the vector length, which seems impossible with this hardcoded length.  And we have the aforementioned case of someone using the same build for regular use and also running tests.  It seems what we should do is remove this hardcoded value and have the api.dis test set the vector length to what it assumes.

Reply:
> And we have the aforementioned case of someone using the same build for regular use and also running tests.

Ah ok, I didn't realise that is a valid use-case. To be clear, you're saying that an SVE user should be able to build with `-DBUILD_TESTS=ON` and specify the vector length which the unit test reference data in `suite/tests/api/dis-a64-sve.txt` should be able to validate against? So, instead of:
```
858057a1 : ldr z1, [x29, #5, mul vl]                : ldr    +0x05(%x29)[32byte] -> %z1
```
with a fixed 256 bits, we have something like:
```
858057a1 : ldr z1, [x29, #5, mul vl]                : ldr    +0x05(%x29)[%sve_vl_bytes] -> %z1
```
where `%sve_vl_bytes` is a runtime user-specified value?

Isn't this overkill for test purposes? With the SVE unit tests we want to check that the vector length is used correctly in calculating number of bytes in memory transfers. The unit tests run with  a **_sample_** length of 32 bytes. Do users want to run unit tests with all possible vector lengths?

Reply:
> what I mean is that someone should be able to build with -DBUILD_TESTS=ON without libdynamorio and libdrdecode suddenly having a hardcoded vector length for all uses outside of tests.

ISWYM. Do you think that's possible if we use a combination of `DR_HOST_NOT_TARGET`, `STANDALONE_DECODER` and `proc_has_feature(FEATURE_SVE)`?

I think this approach is related to the comments raised here https://github.com/DynamoRIO/dynamorio/pull/5835#discussion_r1087301519

Perhaps it's possible to unify the approach to handle both cases.

Reply:
> Perhaps it's possible to unify the approach to handle both cases.

Right, it seems that having an interface to set the target hardware features and calling that from the tests may be the best solution.

Reply:
Done. Added proc_set_feature() and enable_all_test_cpu_features() called from test code to set target hardware features.